hoomd/mpcd/ReverseNonequilibriumShearFlow.cc

```diff
@@ -0,0 +1,327 @@
+    m_hi_pos = make_scalar2(hi_pos - delta_half, hi_pos + delta_half);
+    m_lo_pos = make_scalar2(lo_pos - delta_half, lo_pos + delta_half);
+
+    // explicit check that the two regions for velocity swapping are inside
+    // of the box entirely and don't overlap
+    if (m_hi_pos.y >= global_hi.z)
+        {
+        throw std::runtime_error("slab outside simulation box in ReverseNonequilibriumShearFlow");
+        }
+    if (m_lo_pos.x < global_lo.z)
+        {
+        throw std::runtime_error("slab outside simulation box in ReverseNonequilibriumShearFlow");
+        }
+    if (m_lo_pos.y > m_hi_pos.x)
+        {
+        throw std::runtime_error("slab outside simulation box in ReverseNonequilibriumShearFlow");
+        }
+    }
+
+/*!
+ * \param timestep Current time step of the simulation
+ */
+void mpcd::ReverseNonequilibriumShearFlow::update(uint64_t timestep)
+    {
+    // reset momentum exchange for this step
+    m_momentum_exchange = Scalar(0);
+
+    // don't do anything if this is an empty group
+    const unsigned int n_mpcd = m_mpcd_pdata->getN();
+    if (n_mpcd == 0)
+        return;
+
+    // if slabs have changed, update them
+    if (m_update_slabs)
+        {
+        setSlabs();
+        m_update_slabs = false;
+        }
+
+    findSwapParticles();
+    swapPairMomentum();
+    }
+
+std::vector<std::string> mpcd::ReverseNonequilibriumShearFlow::getProvidedLogQuantities()
+    {
+    std::vector<std::string> ret {"rp_momentum"};
+    return ret;
+    }
+
+/*!
+ * \param quantity Name of the log quantity to get
+ * \param timestep Current time step of the simulation
+ */
+Scalar mpcd::ReverseNonequilibriumShearFlow::getLogValue(const std::string& quantity,
+                                                         uint64_t timestep)
+    {
+    if (quantity == "rp_momentum")
+        {
+        return m_momentum_exchange;
+        }
+    else
+        {
+        throw std::runtime_error("Error getting log value");
+        }
+    return Scalar(0.0);
+    }
+
+/*!
+ * Finds all particles in the "min" and "max" slab in z direction and
+ * puts them into two GPUArrays, sorted by their momentum closest to p_target in x direction.
+ */
+void mpcd::ReverseNonequilibriumShearFlow::findSwapParticles()
+    {
+    // get needed array handles. All in read mode, no particle is changed in this search function
+    ArrayHandle<Scalar4> h_vel(m_mpcd_pdata->getVelocities(),
+                               access_location::host,
+                               access_mode::read);
+    ArrayHandle<Scalar4> h_pos(m_mpcd_pdata->getPositions(),
+                               access_location::host,
+                               access_mode::read);
+
+    // sort pairs according to their sign and momentum
+    // temporary pairs to hold particles in slab
+    // first index is momentum in x dir - second index is particle tag
+    detail::ReverseNonequilibriumShearFlowSorter cmp(m_p_target);
```

Comment:
I think these should actually be two comparators with opposite signs on the target (hi slab looks for velocity closest to -target, lo slab looks for +target in order to swap)
```suggestion
    detail::ReverseNonequilibriumShearFlowSorter cmp_hi(-m_p_target);
    detail::ReverseNonequilibriumShearFlowSorter cmp_lo(m_p_target);
```
The code was working as is, but I am proposing this change so we can also simplify the utility class

Reply:
If I commit these changes, do I need to change anything in the `ReverseNonequilibriumShearFlowUtility.h` file?

Reply:
since that is where this function is coming from

Reply:
Yeah, you would need to apply the changes I suggested below to that file, since giving this signed argument will only work if the implementation of the comparison operators is also updated.